src/http-client/node-http2-client.ts

```diff
@@ -55,143 +120,75 @@ export class NodeHTTP2Client implements HTTPClient {
-      ...DEFAULT_SESSION_OPTIONS,
-      ...options,
-    };
-    // TODO: put a cap on lax idle time
-    this.#idleTime = _options.idleTime;
-    // WIP: should be set to something different for streaming
-    this.#pathName = "/query/1";
+    // typecast safety: we just checked if the entry existed
+    const sessionWrapper = this.#getSessionWrapper() as SessionWrapper;
 
-    try {
-      this.internal = http2.connect(url);
-      this.internal.setTimeout(this.#idleTime, () => {
-        this.close();
-      });
-    } catch (error) {
-      throw new NetworkError(`Could not connect to Fauna`, { cause: error });
+    // if the existing session is closed then a create a new one an update the
+    // shared session map
+    if (sessionWrapper.session.closed) {
+      // cannot reuse sessions once they are closed
+      const new_session = connectNewSession();
+      sessionWrapper.session = new_session;
     }
-  }
 
-  close() {
-    this.internal.close();
+    return sessionWrapper.session;
   }
 
-  async request({
-    data: requestData,
-    headers: requestHeaders,
-    method,
-  }: SessionRequestOptions): Promise<HTTPResponse> {
-    let req: any;
-
-    return new Promise<HTTPResponse>((resolvePromise, rejectPromise) => {
-      const onResponse = (http2ResponseHeaders: any) => {
-        const status = Number(
-          http2ResponseHeaders[http2.constants.HTTP2_HEADER_STATUS]
-        );
-        let responseData = "";
-
-        // append response data to the data string every time we receive new data
-        // chunks in the response
-        req.on("data", (chunk: any) => {
-          responseData += chunk;
-        });
-
-        // Once the response is finished, resolve the promise
-        req.on("end", () => {
-          resolvePromise({
-            status,
-            body: responseData,
-            headers: http2ResponseHeaders,
-          });
-        });
-      };
-
-      try {
-        const httpRequestHeaders: any = {
-          ...requestHeaders,
-          [http2.constants.HTTP2_HEADER_PATH]: this.#pathName,
-          [http2.constants.HTTP2_HEADER_METHOD]: method,
-        };
+  #getSessionWrapper(): SessionWrapper | undefined {
+    return NodeHTTP2Client.#sessionMap.get(this.#sessionKey);
+  }
 
-        req = this.internal
-          .request(httpRequestHeaders)
-          .setEncoding("utf8")
-          .on("error", (error: any) => rejectPromise(error))
-          .on("response", onResponse);
-        req.write(JSON.stringify(requestData), "utf8");
-        req.end();
-      } catch (error) {
-        rejectPromise(error);
-      }
-    });
+  get #sessionKey(): string {
+    // WIP: need to account for streaming
+    return `${this.#url}|${this.#http2_session_idle_ms}`;
   }
 }
+
+/**
+ * Put the session in an object, so we can update the session without deleting
+ * and resetting what is in the Map
+ */
+type SessionWrapper = {
+  session: ClientHttp2Session;
+};
```

Comment:
Now that I have everything working, I don't think this is necessary anymore. But when I naively unwrapped everything to just put a ClientHttp2Session in the static session map, I got unexpected errors from the test to check if sessions are closed. I will need more time to figure this one out we want to drop it. It's not too intrusive as is, though.

Reply:
I'm using this to hold onto references, and renamed to SessionRC in the latest update.